### PR TITLE
LPD-28116 - add permission check to account menu organization selector

### PR DESCRIPTION
**Pull request by wanderlast**
Hi @smottal,

https://liferay.atlassian.net/browse/LPD-28116, tied to LPP.

This basically uses a solution identical to what was implemented in [LPD-24824](https://liferay.atlassian.net/browse/LPD-24824) and applies it to the Account menu's organization selector as well. Sending as a Draft since the ticket is on Fire and needs to be reviewed ASAP.

### CODE REVIEW

**Comment by wanderlast** [comment]
As mentioned in the thread, this is a solution implementing an "Edit Organizations" permission.

The way the solution is meant to work is as follows:
- A user with EDIT organizations permissions can do everything that MANAGE organizations can in the UI but ONLY on organizations they are members of due to the restriction in visibility in AssignableAccountOrganizationSearchContainerFactory.
- A user with MANAGE organizations permissions can do everything that they've always been able to do (nothing has changed for their access level)
- A user with both EDIT and MANAGE organizations permissions works identically to MANAGE Organizations.

Permission checks that were previously against "Manage_Organizations" are now being made against "Edit_Organizations" (see: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountEntryOrganizationRelServiceImpl.java) with a specific check added [here](https://github.com/liferay-commerce/liferay-portal/pull/5025/commits/41a40992554f855bc7223d163732311640bd6abf#diff-866f2eab8371ef46441eb32846fcece661281fc4aeeab08782f7ffd8627759d0R149-R156) to insure that anything checking against Edit Organizations is something that Manage Organizations is permitted to do.

File: modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/dao/search/AssignableAccountOrganizationSearchContainerFactory.java [hunk 1]
```diff
@@ -53,11 +59,26 @@ public static SearchContainer<Organization> create(
...
 				WebKeys.THEME_DISPLAY);
 
+		PermissionChecker permissionChecker =
+			themeDisplay.getPermissionChecker();
+
+		LinkedHashMap<String, Object> params = new LinkedHashMap<>();
+
+		if (!AccountEntryPermission.contains(
+				permissionChecker, accountEntryId,
+				AccountActionKeys.MANAGE_ORGANIZATIONS)) {
```
Comment (smottal) [inline modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/dao/search/AssignableAccountOrganizationSearchContainerFactory.java:69]: I think we have to check also the VIEW permission on organization here

File: modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/security/permission/resource/AccountEntryPermission.java [hunk 2]
```diff
@@ -58,6 +59,22 @@ public static boolean contains(
 		return false;
 	}
 
+	public static boolean hasEditOrManageOrganizationPermission(
```
Comment (smottal) [inline modules/apps/account/account-admin-web/src/main/java/com/liferay/account/admin/web/internal/security/permission/resource/AccountEntryPermission.java:62]: hasEditOrManageOrganizationPermission -> hasEditOrManageOrganization**s**Permission

File: modules/apps/account/account-service/src/main/java/com/liferay/account/internal/security/permission/resource/AccountEntryModelResourcePermission.java [hunk 3]
```diff
@@ -112,8 +112,10 @@ public boolean contains(
 				boolean organizationMember = ArrayUtil.contains(
 					userOrganizationIds, organization.getOrganizationId());
 
-				if (!Objects.equals(
-						actionId, AccountActionKeys.MANAGE_ORGANIZATIONS) &&
+				if ((!Objects.equals(
+						actionId, AccountActionKeys.MANAGE_ORGANIZATIONS) ||
+					 !Objects.equals(
+						 actionId, AccountActionKeys.EDIT_ORGANIZATIONS)) &&
```
Comment (smottal) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/internal/security/permission/resource/AccountEntryModelResourcePermission.java:118]: can you please swipe them to have them alphabetically sorted?

File: modules/apps/account/account-service/src/main/java/com/liferay/account/internal/security/permission/resource/AccountEntryModelResourcePermission.java [hunk 4]
```diff
@@ -144,6 +146,15 @@ public boolean contains(
...
 
+				if (Objects.equals(
+						actionId, AccountActionKeys.EDIT_ORGANIZATIONS) &&
+					permissionChecker.hasPermission(
+						organization.getGroupId(), AccountEntry.class.getName(),
+						accountEntryId, AccountActionKeys.MANAGE_ORGANIZATIONS)) {
+
+					return true;
+				}
+
```
Comment (smottal) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/internal/security/permission/resource/AccountEntryModelResourcePermission.java:157]: can you move this as first check? (Line 114 more or less)

**Review by smottal — CHANGES_REQUESTED** [review]

**Comment by smottal** [comment]
@wanderlast the logic is fine. Just little changes. Thx

**Review by smottal — APPROVED** [review]

**Comment by smottal** [comment]
@wanderlast it's fine!

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

File: modules/test/playwright/tests/account-admin-web/accountPermissions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+	await usersAndOrganizationsPage.goToUsers();
+	await (
+		await usersAndOrganizationsPage.usersTableRowActions(
+			`${user.alternateName}`
+		)
+	).click();
+
+	const pagePromise = context.waitForEvent('page');
+	await usersAndOrganizationsPage.impersonateUserMenuItem.click();
```
Comment (smottal) [inline modules/test/playwright/tests/account-admin-web/accountPermissions.spec.ts:93]: new empty line

File: modules/test/playwright/tests/workspaces/liferay-workspace-marketplace/marketplacePublisherDashboard.spec.ts [hunk 6]
```diff
@@ -52,9 +52,17 @@ test.describe('Can Publish and Manage Solutions', () => {
...
 			await apiHelpers.headlessCommerceAdminCatalog.postCatalog({
 				accountId: account.id,
 			});
+=======
+		await apiHelpers.headlessAdminUser.assignUserToAccountRole(
+			account.id,
+			accountSupplierRole[0].id,
+			user.id
+		);
+>>>>>>> 339768d9a67a5 (LPD-28116 - SF)
```
Comment (smottal) [inline modules/test/playwright/tests/workspaces/liferay-workspace-marketplace/marketplacePublisherDashboard.spec.ts:65]: wrong merge

**Review by smottal — CHANGES_REQUESTED** [review]

**Comment by smottal** [comment]
@wanderlast can you please check the comments and also add two other tests (very similar to this one)?
1) if you have not the EDIT_ORGANIZATIONS but the VIEW_ORGANIZATIONS you see all the orgs
2) if you don't have the EDIT_ORGANIZATIONS and you don't have the VIEW_ORGANIZATIONS the new button is not present

Thx

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
Repushed to fix rebase conflict.

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by smottal** [comment]
ci:test:um

**Review by smottal — APPROVED** [review]

**Comment by smottal** [comment]
ci:test:relevant

**Comment by smottal** [comment]
Hey @wanderlast can you please have a look to the semantic version problem and the failing test? Thx

**Comment by wanderlast** [comment]
@smottal, fixed the versioning and logic so the test should work now.

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:sf

**Review by smottal — APPROVED** [review]

**Comment by smottal** [comment]
ci:test:relevant

**Comment by smottal** [comment]
ci:test:um

**Comment by smottal** [comment]
@wanderlast can you please check the semantic version problem? Also in the CI your playwright tests are failing

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
Had to rebase to resolve merge conflict. I've fixed the Semver and edited the playwright test further.

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by smottal** [comment]
ci:test:relevant

**Comment by smottal** [comment]
ci:forward

**Pull request closed without merging**